src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,808 @@
+"""
+mutable struct MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
+   X::GapObj
+   gens::Vector{<:AbstractMatrixGroupElem}
+   ring::Ring
+   deg::Int
+   descr::Symbol                       # e.g. GL
+   ring_iso::GenRingIso
+   mat_iso::GenMatIso
+
+   function MatrixGroup(m::Int64, F::S) where S<:Ring
+      K = elem_type(S)
+      r = new{K, MatElem{K}}()
+      r.deg = m
+      r.ring = F
+      return r
+   end
+
+   function MatrixGroup(m::Int64, F::S, G_gap::GapObj) where S<:Ring
+      K = elem_type(S)
+      r = new{K, MatElem{K}}()
+      r.deg = m
+      r.ring = F
+      r.X = G_gap
+      return r
+   end
+
+   function MatrixGroup(m::Int64, F::S, L::Vector) where S<:Ring
+      K = elem_type(S)
+      r = new{K, MatElem{K}}()
+      r.deg = m
+      r.ring = F
+      r.gens = L
+      return r
+   end
+end
+
+
+"""
+    MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: GAPGroupElem{MatrixGroup}
+Elements of a group of type `MatrixGroup{RE<:RingElem, T<:MatElem{RE}}`
+"""
+mutable struct MatrixGroupElem{RE<:RingElem, T<:MatElem{RE}} <: AbstractMatrixGroupElem
+   parent::MatrixGroup{RE, T}
+   elm::T
+   X::GapObj
+
+   function MatrixGroupElem(G::MatrixGroup{RE,MatElem{RE}}, x::MatElem{RE}) where RE <: RingElem
+      z = new{RE, MatElem{RE}}()
+      z.parent = G
+      z.elm = x
+      return z
+   end
+
+   function MatrixGroupElem(G::MatrixGroup{RE,MatElem{RE}}, x::MatElem{RE}, x_gap::GapObj) where RE <: RingElem
+      z = new{RE, MatElem{RE}}()
+      z.parent = G
+      z.elm = x
+      z.X = x_gap
+      return z
+   end
+
+   function MatrixGroupElem(G::MatrixGroup{RE,MatElem{RE}}, x_gap::GapObj) where RE <: RingElem
+      z = new{RE, MatElem{RE}}()
+      z.parent = G
+      z.X = x_gap
+      return z
+   end
+end
+
+ring_elem_type(::Type{MatrixGroup{S,T}}) where {S,T} = S
+mat_elem_type(::Type{MatrixGroup{S,T}}) where {S,T} = T
+
+########################################################################
+#
+# Basic
+#
+########################################################################
+
+function Base.show(io::IO, x::MatrixGroup)
+   if isdefined(x, :descr)
+      if x.descr==:GU || x.descr==:SU
+         print(io, string(x.descr), "(",x.deg,",",characteristic(x.ring)^(div(degree(x.ring),2)),")")
+      else
+         print(io, string(x.descr), "(",x.deg,",",order(x.ring),")")
+      end
+   else
+      print(io, "Matrix group of degree ", x.deg, " over ", x.ring)
+   end
+end
+
+function Base.show(io::IO, x::MatrixGroupElem)          #TODO: is this correct?
+   if isdefined(x, :elm)
+      display(x.elm)
+   else
+      print(io, GAP.gap_to_julia(GAP.Globals.StringViewObj(x.X)))
+   end
+end
+
+group_element(G::MatrixGroup, x::GapObj) = MatrixGroupElem(G,x)
+
+function Base.getproperty(G::MatrixGroup, sym::Symbol)
+
+   if sym === :mat_iso || sym === :ring_iso
+      if isdefined(G,sym) return getfield(G,sym) end
+      fm = gen_mat_iso(G.deg, G.ring)
+      setfield!(G,:ring_iso,fm.fr)
```

Comment:
Why not `G.ring_iso = fm.fr` ? `setfield!` is intended to be used when you have the name of the field in a variable, whose values is not knows statically.

Reply:
I did not use `G.ring_iso = fm.fr` because I was not sure to create infinite recursions. If it works, would it be sensible to change it?

Reply:
Oh I didn't think about that... is `setproperty!` overloaded somewhere?

Reply:
But yes, if possible, `G.ring_iso  =...` is much better than `setfield(G, :ring_iso, ...)`.